File: Backend/app/schemas/calculator.py
"""Schemas for the calculate_damage tool. See Docs/backend/damage-calc.md for
the exact formula scope (what's implemented vs. explicitly deferred)."""

from typing import Literal

from pydantic import BaseModel, Field

Status = Literal["brn", "psn", "tox", "par", "slp", "frz"]


class PokemonBattleState(BaseModel):
    """A Pokemon set as used in a single damage calculation — deliberately
    lighter-weight than the Team Builder's PokemonSet (see app/schemas/team.py),
    since a calculator query doesn't need moves other than the one being used."""

    species_id: str
    level: int = 100
    nature: str = "hardy"
    evs: dict[str, int] = Field(default_factory=dict)
    """Missing stat keys default to 0. Not strictly validated against the real
    508 total / 252 per-stat caps — see Docs/backend/damage-calc.md."""
    ivs: dict[str, int] = Field(default_factory=dict)
    """Missing stat keys default to 31 (perfect IVs)."""
    ability: str | None = None
    """Defaults to the species' first ability slot ("0") if not given."""
    item: str | None = None
    status: Status | None = None
    tera_type: str | None = None
    stat_stages: dict[str, int] = Field(default_factory=dict)
    """atk/def/spa/spd/spe -> -6..6. HP is not a valid key (HP has no stages)."""
    current_hp_percent: float = 100.0
    """For KO-chance context — e.g. "can this already-damaged Pokemon finish
    off a target at 40% HP." Defaults to full HP."""


class FieldConditions(BaseModel):
    weather: Literal["sun", "rain"] | None = None
    reflect: bool = False
    light_screen: bool = False
    aurora_veil: bool = False
    is_critical: bool = False
    spread_move: bool = False
    """The 0.75x reduction for a move hitting multiple targets in a doubles
    battle — a manual flag rather than full VGC doubles-targeting simulation,
    see Docs/backend/damage-calc.md."""


class DamageCalcRequest(BaseModel):
    attacker: PokemonBattleState
    defender: PokemonBattleState
    move_id: str
    field: FieldConditions = Field(default_factory=FieldConditions)


class DamageCalcResult(BaseModel):
    move_name: str
    move_type: str
    category: str
    is_immune: bool
    type_effectiveness: float
    stab_multiplier: float
    rolls: list[int]
    """The 16 damage values for the 85%-100% random roll, in raw HP."""
    min_damage: int
    max_damage: int
    min_percent: float
    max_percent: float
    defender_max_hp: int
    ko_chance_description: str

File: Frontend/src/types/calculator.ts
export type Status = "brn" | "psn" | "tox" | "par" | "slp" | "frz";

export type PokemonBattleState = {
  species_id: string;
  level: number;
  nature: string;
  evs: Record<string, number>;
  ivs: Record<string, number>;
  ability?: string | null;
  item?: string | null;
  status?: Status | null;
  tera_type?: string | null;
  stat_stages: Record<string, number>;
  current_hp_percent: number;
};

export type FieldConditions = {
  weather?: "sun" | "rain" | null;
  reflect: boolean;
  light_screen: boolean;
  aurora_veil: boolean;
  is_critical: boolean;
  spread_move: boolean;
};

export type DamageCalcRequest = {
  attacker: PokemonBattleState;
  defender: PokemonBattleState;
  move_id: string;
  field: FieldConditions;
};

export type DamageCalcResult = {
  move_name: string;
  move_type: string;
  category: string;
  is_immune: boolean;
  type_effectiveness: number;
  stab_multiplier: number;
  rolls: number[];
  min_damage: number;
  max_damage: number;
  min_percent: number;
  max_percent: number;
  defender_max_hp: number;
  ko_chance_description: string;
};

export function defaultBattleState(): PokemonBattleState {
  return {
    species_id: "",
    level: 100,
    nature: "hardy",
    evs: {},
    ivs: { hp: 31, atk: 31, def: 31, spa: 31, spd: 31, spe: 31 },
    ability: null,
    item: null,
    status: null,
    tera_type: null,
    stat_stages: {},
    current_hp_percent: 100,
  };
}

export function defaultFieldConditions(): FieldConditions {
  return {
    weather: null,
    reflect: false,
    light_screen: false,
    aurora_veil: false,
    is_critical: false,
    spread_move: false,
  };
}
